Xplat/src/main/java/vazkii/botania/common/item/AstrolabeItem.java

```diff
@@ -110,22 +116,28 @@ public boolean placeAllBlocks(ItemStack stack, Player player) {
 
 		ItemStack stackToPlace = new ItemStack(getBlock(stack));
 		for (BlockPos coords : blocksToPlace) {
-			placeBlockAndConsume(player, stack, stackToPlace, coords);
+			BlockPlaceContext placeContext = BlockPlaceContext.at(ctx, coords, ctx.getClickedFace());
+			placeBlockAndConsume(player, stack, stackToPlace, placeContext);
 		}
 		ManaItemHandler.instance().requestManaExact(stack, player, cost, true);
 
 		return true;
 	}
 
-	private void placeBlockAndConsume(Player player, ItemStack requestor, ItemStack blockToPlace, BlockPos coords) {
+	private void placeBlockAndConsume(Player player, ItemStack requestor, ItemStack blockToPlace, BlockPlaceContext ctx) {
 		if (blockToPlace.isEmpty()) {
 			return;
 		}
 
 		Block block = Block.byItem(blockToPlace.getItem());
-		BlockState state = block.defaultBlockState();
+		BlockState state = block.getStateForPlacement(ctx);
+		if (state == null) {
+			return;
+		}
+		BlockPos coords = ctx.getClickedPos();
 		player.level.setBlockAndUpdate(coords, state);
 		player.level.levelEvent(LevelEvent.PARTICLES_DESTROY_BLOCK, coords, Block.getId(state));
+		player.level.gameEvent(GameEvent.BLOCK_PLACE, coords, GameEvent.Context.of(player, state));
```

Comment:
hmmm, reading the code here I'm a bit surprised it wasn't migrated to substituteUse. Basically a couple years back I replaced most of the block-placing code in Botania to do a "swap into hand, call vanilla use, swap back", using `PlayerHelper.substituteUse`. I'm guessing I missed this spot.

Ideally, we should be using that here too, because then vanilla code will handle all the post-place logic, firing game events, etc.